pkg/kubectl/cmd/set/set_env.go

```diff
@@ -195,23 +251,43 @@ func (o *EnvOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, args []stri
 	o.Output = cmdutil.GetFlagString(cmd, "output")
 	o.From = cmdutil.GetFlagString(cmd, "from")
 	o.Prefix = cmdutil.GetFlagString(cmd, "prefix")
+	o.Keys = cmdutil.GetFlagStringSlice(cmd, "keys")
 	o.DryRun = cmdutil.GetDryRunFlag(cmd)
 
-	o.EnvArgs = envArgs
-	o.Resources = resources
 	o.Cmd = cmd
 
 	o.ShortOutput = cmdutil.GetFlagString(cmd, "output") == "name"
 
+}
+
+func (o *EnvOptions) Validate(args []string) error {
+	if o.All && len(o.Selector) > 0 {
+		return fmt.Errorf("cannot set --all and --selector at the same time")
+	}
+	resources, envArgs, ok := envutil.SplitEnvironmentFromResources(args)
+	if !ok {
+		return cmdutil.UsageErrorf(o.Cmd, "all resources must be specified before environment changes: %s", strings.Join(args, " "))
+	}
+
+	if len(o.Filenames) == 0 && len(resources) == 0 {
+		return cmdutil.UsageErrorf(o.Cmd, "one or more resources must be specified as <resource> <name> or <resource>/<name>")
+	}
+
 	if o.List && len(o.Output) > 0 {
 		return cmdutil.UsageErrorf(o.Cmd, "--list and --output may not be specified together")
 	}
 
+	if len(o.Keys) > 0 && len(o.From) == 0 {
+		return cmdutil.UsageErrorf(o.Cmd, "when specifying --keys, a configmap or secret must be provided with --from")
+	}
+
+	o.EnvArgs = envArgs
+	o.Resources = resources
 	return nil
 }
 
 // RunEnv contains all the necessary functionality for the OpenShift cli env command
-func (o *EnvOptions) RunEnv(f cmdutil.Factory) error {
+func (o *EnvOptions) RunEnv(f cmdutil.Factory, configMaps []*v1.ConfigMap, secrets []*v1.Secret) error {
```

Comment:
These should not be args passed to RunEnv. Aggregate them on the options struct

Reply:
We also should not be passing the factory as an arg here, but since that is pre-existing logic, can be addressed in a separate PR.

Reply:
> "We also should not be passing the factory as an arg here, but since that is pre-existing logic, can be addressed in a separate PR."

Sure, the current testing seems to be dependent on the ability to inject this factory though unless I'm overlooking something?

Reply:
Usually we take the bits we need from the factory (any methods) and tack them on to the options struct in `Complete()`: `o.ClientSet = f.KubernetesClientSet` (builder logic also usually takes place in Complete, with the resulting infos being aggregated)

Reply:
Great, ok I can take a look at doing that in separate pr or commit

Reply:
A separate pr sounds great for this. Thanks